File: 4_5/p047.py
import sys
sys.setrecursionlimit(1000000)

n, m = map(int, input().split())
# adjacency list
al = [list() for _ in range(n+1)]
g1 = set() # グループ1
g2 = set() # グループ2
for _ in range(m):
    a, b = map(int, input().split())
    al[a].append(b)
    al[b].append(a)
vis = [False]*(n+1)
vis[0] = True
def dfs(pos, g_flg):
    for i in al[pos]:
        if g_flg:
            g1.add(i)
        else:
            g2.add(i)
        if not vis[i]:
            vis[i] = True
            dfs(i, not g_flg)
flg = True

for i in range(len(vis)):
    if not vis[i]:
        dfs(i, True)
        if len(g1) != len(g1-g2):
            flg = False
            break
        g1 = set()
        g2 = set()

if flg:
    print('Yes')
else:
    print('No')


###########################################
import sys

sys.setrecursionlimit(210000)

def dfs(pos):
    for i in g[pos]:
        if color[i] == 0:
            color[i] = 3-color[pos]
            dfs(i)

n, m = map(int, input().split())
g = [list() for _ in range(n+1)]
al = [-1]*(m)
bl = [-1]*(m)
for i in range(m):
    al[i], bl[i] = map(int, input().split())
    g[al[i]].append(bl[i])
    g[bl[i]].append(al[i])

color = [0]*(n+1)

for i in range(1, n+1):
    if color[i] == 0:
        color[i] = 1
        dfs(i)
ans = True
for i in range(1, n+1):
    if color[al[i]] == color[bl[i]]:
        ans = False
        break
if ans:
    print('Yes')
else:
    print('No')

File: 4_5/p047.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 10000000
)

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
	scan.Split(bufio.ScanWords)
}

var seen []int
var G [][]int

func main() {
	n := NumStdin()
	m := NumStdin()

	G = make([][]int, n)
	for i := 0; i < m; i++ {
		u, v := NumStdin(), NumStdin()
		G[u-1] = append(G[u-1], v-1)
		G[v-1] = append(G[v-1], u-1)
	}

	seen = make([]int, n)
	for i := 0; i < n; i++ {
		seen[i] = -1
	}

	res := false
	for i := 0; i < n; i++ {
		if seen[i] != -1{
			continue
		}
		res = dfs(i, -1)
		if !res{
			break
		}
	}

	if res {
		fmt.Println("Yes")
	}else{
		fmt.Println("No")
	}
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}

func dfs(u, flg int) bool {
	if flg == -1{
		flg = 0
	}
	seen[u] = flg

	ret := true
	for _, v := range G[u] {
		if seen[u] == seen[v]{
			ret = false
			break
		}else if seen[v] == -1{
			ret = dfs(v, 1-flg)
			if !ret{
				break
			}
		}
	}

	return ret
}
